### DISCUSSION TITLE
Is there a way to pin PDM version in pyproject.toml?

### DISCUSSION THREAD

**Discussion by netrix** [pdm-project/pdm#2332] [Q&A]
I have situation where I use PDM for my Python project. This project is hosted in Gitlab and has pipeline where I use a docker image with PDM (along with `tox-pdm`) already installed with version 2.2 (quite old).

Recently I had a situation where I need to update lock file (I've added some dependency) and I used newer version od PDM for it. After pushing changes to Gitlab the pipeline stopped working because it couldn't find my internally hosted packages (I've added `tool.pdm.source` to `pyproject.toml`). It looks like the lock file definition changed between 2.2 and 2.9 (that I've used to generate new lock file).

Is there a way to lock to the specific version of `pdm` in a way that `pdm` itself will throw an error if the version is not within constraints ?

**Comment by frostming** [comment]
This is the purpose of the lockfile version, but for some reasons, the version is defined badly and compatibility has not been fully guaranteed. We don't need another version.
